pkg/api/steve/collections/collections.go

```diff
@@ -0,0 +1,699 @@
+}
+
+// hasAccess checks the auth headers set by the kube API server for the user's access to the requested resource in the requested project.
+func (h *handler) hasAccess(r *http.Request, namespace string, resource schema.GroupVersionResource) bool {
+	var groups []string
+	if g, ok := r.Header["X-Remote-Group"]; ok {
+		groups = g
+	}
+	review := authzv1.SubjectAccessReview{
+		Spec: authzv1.SubjectAccessReviewSpec{
+			User:   r.Header.Get("X-Remote-User"),
+			Groups: groups,
+			ResourceAttributes: &authzv1.ResourceAttributes{
+				Verb:      "list",
+				Resource:  resource.Resource,
+				Group:     resource.Group,
+				Namespace: namespace,
+			},
+		},
+	}
+	result, err := h.sarClient.Create(r.Context(), &review, metav1.CreateOptions{})
+	return err == nil && result.Status.Allowed
+}
+
+// projectsAndNamespaces sorts a list of projects and namespaces specified by 'fieldSelector=prjorns=...'
+// into separate lists of projects and namespaces.
+func (h *handler) projectsAndNamespaces(projectsOrNamespaces []string) ([]string, map[string]*corev1.Namespace, error) {
+	projects := []string{}
+	namespaces := map[string]*corev1.Namespace{}
+	for _, p := range projectsOrNamespaces {
+		ns, err := h.namespaceCache.Get(p)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return nil, nil, fmt.Errorf("failed to look up namespace %s in cache: %w", p, err)
+		}
+		if ns != nil {
+			_, ok := ns.GetLabels()[ParentLabel]
+			if ok {
+				projects = append(projects, ns.Name)
+				continue
+			}
+			namespaces[ns.Name] = ns
+		}
+	}
+	return projects, namespaces, nil
+}
+
+func sortCollection(resourceCollection []unstructured.Unstructured) {
+	sort.Slice(resourceCollection, func(i, j int) bool {
+		objI := resourceCollection[i]
+		objJ := resourceCollection[j]
+		if objI.GetNamespace() < objJ.GetNamespace() {
+			return true
+		}
+		if objI.GetNamespace() > objJ.GetNamespace() {
+			return false
+		}
+		return objI.GetName() < objJ.GetName()
+	})
+}
+
+func emptyResourceVersion(ctx context.Context, gvr schema.GroupVersionResource, resourceClient dynamic.ResourceInterface) (string, error) {
+	// get the list but ignore the resources, this is needed to get the list resource version
+	resourceList, err := resourceClient.List(ctx, metav1.ListOptions{Limit: 1})
+	if err != nil {
+		return "", fmt.Errorf("failed to get resource version for resource %s: %w", gvr.Resource, err)
+	}
+	return resourceList.GetResourceVersion(), nil
+}
+
+func responseData(resource schema.GroupVersionResource, kind, resourceVersion string, items []unstructured.Unstructured) map[string]interface{} {
+	return map[string]interface{}{
+		"apiVersion": resource.GroupVersion().String(),
+		"kind":       kind,
+		"metadata": map[string]interface{}{
+			"resourceVersion": resourceVersion,
+		},
+		"items": items,
+	}
+}
+
+func returnResp(w http.ResponseWriter, resp map[string]interface{}) {
+	resourceJSON, err := json.Marshal(resp)
```

Comment:
I think it will be safer to use an `unstructuredList` than defining our own map
```suggestion
func responseData(resource schema.GroupVersionResource, kind, resourceVersion string, items []unstructured.Unstructured) unstructured.UnstructuredList{
	 list := unstructured.UnstructuredList {
		Items: items,
	 }
	 list.SetAPIVersion(resource.GroupVersion().String())
	 list.SetKind(kind)
	 list.SetResourceVersion(resourceVersion)
	
	return list
}

func returnResp(w http.ResponseWriter, resp unstructured.UnstructuredList) {
	resourceJSON, err := json.Marshal(resp)
```